Homework №7/Tree/Tree/Tree.c

```diff
@@ -0,0 +1,394 @@
+    }
+    if ((x->parent->leftSon == x && x->parent->parent->leftSon == x->parent)
+        || (x->parent->rightSon == x && x->parent->parent->rightSon == x->parent))
+    {
+        zigZig(x);
+    }
+    else
+    {
+        zigZag(x);
+    }
+    return splay(x);
+}
+
+
+Node* addNode(Node* root, int key, const char* value, Error* error)
+{
+    if (*error != NOT_ERROR)
+    {
+        return root;
+    }
+    char* copyValue = calloc(strlen(value) + 1, sizeof(char));
+    if (copyValue == NULL)
+    {
+        *error = INSUFFICIENT_MEMORY;
+        return root;
+    }
+    strcpy(copyValue, value);
+    if (root == NULL)
+    {
+        Node* newRoot = (Node*)calloc(1, sizeof(Node));
+        if (newRoot == NULL)
+        {
+            *error = INSUFFICIENT_MEMORY;
+            free(copyValue);
+            return root;
+        }
+        newRoot->key = key;
+        newRoot->value = copyValue;
+        return newRoot;
+    }
+    Node* currentRoot = root;
+    while (currentRoot != NULL)
+    {
+        if (key > currentRoot->key)
+        {
+            if (currentRoot->rightSon == NULL)
+            {
+                Node* newRoot = (Node*)calloc(1, sizeof(Node));
+                if (newRoot == NULL)
+                {
+                    *error = INSUFFICIENT_MEMORY;
+                    free(copyValue);
+                    return root;
+                }
+                newRoot->key = key;
+                newRoot->value = copyValue;
+                attach(currentRoot, newRoot, right);
+                return splay(newRoot);
+            }
+            currentRoot = currentRoot->rightSon;
+        }
+        else if (key == currentRoot->key)
+        {
+            free(currentRoot->value);
+            currentRoot->value = copyValue;
+            return splay(currentRoot);
+        }
+        else
+        {
+            if (currentRoot->leftSon == NULL)
+            {
+                Node* newRoot = (Node*)calloc(1, sizeof(Node));
+                if (newRoot == NULL)
+                {
+                    *error = INSUFFICIENT_MEMORY;
+                    free(copyValue);
+                    return root;
+                }
+                newRoot->key = key;
+                newRoot->value = copyValue;
+                attach(currentRoot, newRoot, left);
+                return splay(newRoot);
+            }
+            currentRoot = currentRoot->leftSon;
+        }
+    }
+    free(copyValue);
+    return root;
+}
+
+Node* search(Node* root, int key)
+{
+    Node* newRoot = root;
+    while (newRoot != NULL)
+    {
+        if (key > newRoot->key)
+        {
+            newRoot = newRoot->rightSon;
+        }
+        else if (key < newRoot->key)
+        {
+            newRoot = newRoot->leftSon;
+        }
+        else
+        {
+            return newRoot;
+        }
+    }
+    return NULL;
+}
+
+bool inTree(Node* root, int key)
+{
+    Node* searchResult = search(root, key);
+    return searchResult != NULL;
```

Comment:
Можно было просто 
```suggestion
    return search(root, key) != NULL;
```
Хотя Ваш вариант немного удобнее для отладки.